pandaserver/dataservice/finisher.py

```diff
@@ -0,0 +1,304 @@
+        """
+        Constructs all the necessary attributes for the Finisher object.
+
+        Parameters
+        ----------
+            taskBuffer : TaskBuffer
+                The task buffer that contains the jobs
+            dataset : DatasetSpec
+                The dataset to be transferred
+            job : JobSpec, optional
+                The job to be transferred (default is None)
+            site : str, optional
+                The site where the job is to be transferred (default is None)
+        """
+        threading.Thread.__init__(self)
+        self.dataset = dataset
+        self.task_buffer = taskBuffer
+        self.job = job
+        self.site = site
+
+    def create_metadata_node(self, doc: Document, att_name: str, att_value: str):
+        """
+        Create a metadata node for the XML document.
+
+        Parameters:
+        doc (xml.dom.minidom.Document): The XML document.
+        att_name (str): The attribute name.
+        att_value (str): The attribute value.
+
+        Returns:
+        xml.dom.minidom.Element: The created metadata node.
+        """
+        metadata_node = doc.createElement("metadata")
+        metadata_node.setAttribute("att_name", att_name)
+        metadata_node.setAttribute("att_value", str(att_value))
+        return metadata_node
+
+    def create_file_node(self, doc: Document, file, failed_files: List[str],
+                         no_out_files: List[str]):
+        """
+        This function creates a file node for the XML document.
+
+        Parameters:
+        doc (xml.dom.minidom.Document): The XML document.
+        file (FileSpec): The file specification.
+        failed_files (list): List of failed files.
+        no_out_files (list): List of files with no output.
+
+        Returns:
+        xml.dom.minidom.Element: The created file node or None if the file is in the list of failed files or no output files.
+        """
+        if file.lfn in failed_files + no_out_files:
+            return None
+        file_node = doc.createElement("File")
+        file_node.setAttribute("ID", file.GUID)
+        log_node = doc.createElement("logical")
+        lfn_node = doc.createElement("lfn")
+        lfn_node.setAttribute("name", file.lfn)
+        log_node.appendChild(lfn_node)
+        file_node.appendChild(log_node)
+        file_node.appendChild(self.create_metadata_node(doc, "fsize", file.fsize))
+        if file.checksum.startswith("ad:"):
+            file_node.appendChild(self.create_metadata_node(doc, "adler32",
+                                                            re.sub("^ad:", "", file.checksum)))
+        else:
+            file_node.appendChild(self.create_metadata_node(doc, "md5sum",
+                                                            re.sub("^md5:", "", file.checksum)))
+        return file_node
+
+    def create_xml_doc(self, job, failed_files: List[str], no_out_files: List[str]):
```

Comment:
Where is the XML doc going? Is it possible to change this to json? (If the XML is consumed externally or we prefer to keep this homogeneous just ignore the comment, but XML seems a bit complex)

Reply:
I had the impression that it is used externally somehow, @tmaeno could you please comment?

Reply:
The doc is used by Adder to add files to _tid datasets. Adder can work both with json and xml.

Reply:
Ok thank you, then it would be better to change it to json (also for readability, less complexity and etc). I will work on it tomorrow.